### ISSUE TITLE
Implement canvasproperties glade file

### ISSUE THREAD

**Issue by Keyikedalube** [synfig/synfig#1552]
**Synfig version & platform**:
Building from master branch
Debian sid

**Issue description**:
To ease UI designers, it's wise to separate the handwritten UI code from canvasproperties.cpp to its own glade (UI) file
And implement only the logic inside the .cpp file

That way it'll be much easier to focus only on the design aspects without having to refer to GTKMM API docs all the time

**Comment by Keyikedalube** [comment]
I'm experimenting/working on this issue ;)

**Comment by rodolforg** [comment]
@Keyikedalube Great :D

Synfig Custom widgets should support GtkBuilder (it should be Gtk::Buildable). I made something here 025eec22 for Widget_Time - this allow us to refer to them on Glade too (but with a placeholder - not the real appearance)

I converted a dialog to UI here 9207c07d .

Ideally, we would have a library (.so/dll/dynlib) for collecting custom widgets to be able to preview better the dialogs on Glade.
I don't know if it would be possible as they're in C++ though.

**Comment by Keyikedalube** [comment]
I'm trying to separate GammaPattern from [renddesc.cpp](https://github.com/synfig/synfig/blob/a74530f8ed068502a6940ade6efa6d57afe943fa/synfig-studio/src/gui/renddesc.cpp#L84) to its own source file but I'm stuck with this error
> gammapattern.h:13:2: error: ‘Gamma’ does not name a type; did you mean ‘gamma’?

I believe this include is the correct one to successfully build GammaPattern class

https://github.com/synfig/synfig/blob/a74530f8ed068502a6940ade6efa6d57afe943fa/synfig-studio/src/gui/renddesc.h#L40

Here's how I implemented the gammapattern.h file
```c++
#ifndef __SYNFIG_GTKMM_GAMMAPATTERN_H
#define __SYNFIG_GTKMM_GAMMAPATTERN_H

#include <gtkmm/drawingarea.h>

#include <synfig/renddesc.h>

namespace studio {

class GammaPattern : public Gtk::DrawingArea
{
	Gamma gamma;
	int tile_w, tile_h, gradient_h;
	Cairo::RefPtr<Cairo::SurfacePattern> pattern;
	
public:
	GammaPattern();

	const set_gamma(const Gamma &);
	const Gamma &get_gamma();

	virtual bool on_draw(const Cairo::RefPtr<Cairo::Context> &);
};

};

#endif
```

I'm using CMake to build synfig for now. I've included the definition file in target_sources(synfigstudio ...)
What am I missing?

**Comment by ice0** [comment]
You need to use `synfig::Gamma`, it added indirectly via `using namespace synfig;`

**Comment by Keyikedalube** [comment]
@ice0 Thanks! How stupid of me. I did not notice that I inserted `using namespace synfig` in the .cpp implementation file and I missed the error message indicating .h
I was too tired and a bit desperate ;P

Anyway, so much thanks for the pointer!

**Comment by rodolforg** [comment]
Why are you including `synfig/renddesc.h` instead of `synfig/gamma.h`?

**Comment by Keyikedalube** [comment]
I'm duplicating Widget_RendDesc class to CanvasProperties, so including `synfig/renddesc.h` is a necessity

Because the only way I can implement the UI file is to re-implement all the same layouts from Widget_RendDesc and afterward pull in the needed widget variables from the UI file for CanvasProperties
Then safely deprecate Widget_RendDesc class for future  #721

Thought about implementing glade file for Widget_RendDesc but it's not intuitive because then there would be no canvasproperties UI file for reason of including Widget_RendDesc UI as a header file
Most UI components for CanvasProperties are already in Windget_RendDesc
So restructuring is needed here

**Comment by Keyikedalube** [comment]
Thanks @rodolforg :)

Had to use `synfig/color/gamma.h` though
make couldn't find `synfig/gamma.h`

**Comment by Keyikedalube** [comment]
https://github.com/synfig/synfig/blob/a74530f8ed068502a6940ade6efa6d57afe943fa/synfig-studio/src/gui/dialogs/canvasproperties.cpp#L93-L102

Is the `idLabel` & `entry_id` output for the above code like this? (But hidden)
![Screenshot from 2020-07-16 17-29-46](https://user-images.githubusercontent.com/24214509/87672822-2b5e3100-c791-11ea-8005-2f82c57fa2c0.png)

Because I haven't seen an actual dialog with that idLabel and entry_id on it...

**Comment by rodolforg** [comment]
As far as I understood reading the code, this field is for rendering exported canvas.

However,  I don't know a GUI way to call this dialog for those canvases

**Comment by rodolforg** [comment]
Ok, I got it.

You must open the exported canvas via Canvas Navigator panel. :)

You can export canvas by exporting value of the "canvas" (¬¬) parameter of a Group Layer, for example.

**Comment by Keyikedalube** [comment]
Thanks! Tested it and got ID output in the top :)
![Screenshot from 2020-07-17 07-20-51](https://user-images.githubusercontent.com/24214509/87739784-487d1900-c7fe-11ea-912d-cd105c3a9a2b.png)

**Comment by Keyikedalube** [comment]
Implementing Widget_Link as custom widget
https://github.com/synfig/synfig/blob/master/synfig-studio/src/gui/widgets/widget_link.cpp
```c++
Glib::ObjectBase
*Widget_Link::wrap_new(
		GObject *o,
		const string &tlt_inactive,
		const string &tlt_active)
{
	if (gtk_widget_is_toplevel(GTK_TOGGLE_BUTTON(o)))
		return new Widget_Link(GTK_TOGGLE_BUTTON(o), tlt_inactive, tlt_active);
	else
		return Gtk::manage(new Widget_Link(GTK_TOGGLE_BUTTON(o), tlt_inactive, tlt_active));
}
```

I get this error for if block:
> cannot convert ‘GtkToggleButton*’ {aka ‘_GtkToggleButton*’} to ‘GtkWidget*’ {aka ‘_GtkWidget*’}

Also, the constructor for Widget_Link takes two std::string arguments, I don't really know how to pass those args to Glib::wrap part
```c++
Glib::wrap_register(gtype, Widget_Link::wrap);
```

**Comment by Keyikedalube** [comment]
I think there's no way to send tooltip message string when we try to `widget_link->get_widget("widget_name", widget)` from glade file

Instead, we would have to set the tooltip message directly on the glade file using
```XML
<object ...>
  <property name="tooltip_text">Click to link/unlink</property>
</object>
```
I guess the only downside to this would be static message instead of the dynamic one synfig has right now

**Comment by rodolforg** [comment]
AFAIK, `WrapNewFunction` passed in `void Glib::wrap_register(GType type, WrapNewFunction func)` takes the allocated C object into a new C++ object, calling the `YourClass(BaseObjectType* cobject)` constructor (I'm calling it C-wrap constructor).

In this case, `BaseObjectType` is `GtkToggleButton`, that doesn't take those parameters in current single C++ constructor of `Widget_Link`.

Workaround here is easy: create a default constructor (without any parameters): it's "just" tooltips anyway.
So the C-wrap constructor can work normally.
Then, via glade/gtkbuilder properties, you load the custom tooltips. I don't remember know how to do it, if I did it before XD

---
Yes, it's very poorly documented.

Here's some related info around the web:

- https://stackoverflow.com/questions/10826435/how-do-i-add-a-custom-gtkmm-widget-to-glade
- https://web.archive.org/web/20160729175637/https://szilard.blogs.balabit.com/en/2012/06/how-to-add-custom-gtkmm-widget-to-glade/
- https://developer.gnome.org/gtkmm-tutorial/stable/sec-basics-gobj-and-wrap.html.en
- https://developer.gnome.org/gtkmm-tutorial/stable/sec-custom-widgets.html.en
- http://gtk.10911.n7.nabble.com/Setting-custom-Glib-Property-values-in-Gtk-Builder-ui-XML-td95393.html

**Comment by rodolforg** [comment]
I believe it's here: https://developer.gnome.org/gtkmm-tutorial/stable/sec-builder-using-derived-widgets.html.en

**Comment by rodolforg** [comment]
@Keyikedalube that's way: you'll implement Glib::Property ;)


``` c++
// .h
#include <glibmm/property.h>

private:
	Glib::Property<std::string> property_mine_;
public:
	Glib::PropertyProxy<std::string> property_mine() { return property_mine_.get_proxy();}
```

``` c++
// .cpp
Widget_X::Widget_X() :
	Glib::ObjectBase("widget_x"), // name in glade: gtkmm__CustomObject_widget_x
	property_mine_(*this, "property_name_for_glade", "My default") // third argument is optional
{
	init();
}

Widget_X::Widget_X(BaseObjectType* cobject) :
	Glib::ObjectBase("widget_x"), // name in glade: gtkmm__CustomObject_widget_x
	Gtk::Entry(cobject),
	property_mine_(*this, "property_name_for_glade", "My default") // third argument is optional
{
	init();
}

void
Widget_X::init()
{
	set_tooltip_text(property_mine_.get_value());
}
```

**Comment by Keyikedalube** [comment]
About 80% done...

Glade file is too ugly now with lots of custom widgets
![Peek 2020-07-19 20-28](https://user-images.githubusercontent.com/24214509/87878010-38805780-c9ff-11ea-8615-72b3ab49cf3d.gif)

Derived GtkToggleButton implemented in synfig ie., Widget_Link is too simple
No special added feature at all
I feel it's better adding the standard GtkToggleButton from Glade options and programmatically manipulating the icons
Haven't tried it yet

And here's how the loaded glade dialog looks after building and running
![Screenshot from 2020-07-19 20-25-07](https://user-images.githubusercontent.com/24214509/87878142-1affbd80-ca00-11ea-845f-d6d846f36db7.png)

**Comment by rodolforg** [comment]
Did you try to add the custom widgets in synfig glade catalog xml?

**Comment by Keyikedalube** [comment]
Yes, so far `gammapattern` and `widget_link` have been implemented as custom widgets

**Comment by rodolforg** [comment]
Did you see that notebook container is a separated widget? It's also used in other dialogs, like Render.

**Comment by Keyikedalube** [comment]
Yes, I noticed that

For CanvasProperties I'm duplicating the notebook container all over again
That way I don't have to modify three separate classes which would be tedious

The only modification made to Widget_Render is separating out GammaPattern to its own file so both Widget_Render and CanvasProperties class can instantiate GammaPattern object by including it as a header file
One invokes it with pure C++ code and the other invokes it as a custom widget inside the glade file  

Once this issue is done, Widget_Render (notebook container) can be safely deprecated and new Render dialog #721 can be implemented in the future

**Comment by Keyikedalube** [comment]
Implementation is done but I haven't modified the build script for Autotools yet

On CMake, a new build target is implemented called `build_resource`, this time our resource (glade file) is generated to a C file using glib-compile-resources tool and later compiled with synfigstudio. That way we don't have to ship the .glade file because the UI is loaded as library.

There is a bug (and maybe more) with the newly implemented CanvasPorperties class. I'm working on fixing it before submitting the PR

PS: For compiling gresource on Autotools, I might ditch it if the above pending issue persists. It's new to me :(

**Comment by Keyikedalube** [comment]
Made the PR

Couldn't solve the above post-issue... More info on the PR

**Comment by Keyikedalube** [comment]
Behavior issue solved in commit d1ebfd83a56d567a20758fdd17cdcfae6185ad69
Whew..!

Need help with updating the Autotools build script

**Comment by rodolforg** [comment]
> Need help with updating the Autotools build script

I can help with that, but not right now…

**Comment by Keyikedalube** [comment]
Closed in favor of #777

*Reason discussed [here](https://github.com/synfig/synfig/pull/1577#issuecomment-719881157) and [here](https://github.com/synfig/synfig/pull/1577#issuecomment-720063687)*
